### PR TITLE
Use retworkx

### PR DESCRIPTION
**Pull request by mtreinish**
### Summary

This PR switches the graph library used for the DAGCircuit class to use retworkx[1] instead of networkx. This provides a significant speed up over networkx because retworkx is written in rust instead of pure python.

[1] https://github.com/mtreinish/retworkx

### Details and comments

Co-Authored-By: Lauren Capelluto <laurencapelluto@gmail.com>

### CODE REVIEW

**Comment by ajavadia** [comment]
I can see #2860 uses NetworkX `max_weight_matching` and `floyd_warshall` algorithms. Do you have any sense of how easy these are to implement?

#3762 (which is a restricted version of the above) also uses NetworkX's `dfs_edges`, `floyd_warshall_numpy`, `find_cycle`. Plus a bunch of other methods for building simple graphs for testing. Can it continue to use NetworkX (potentially taking the performance hit)?

**Comment by mtreinish** [comment]
I'll have to look into those algorithms to see how they work, but none of the algorithms I've looked at so far have been difficult to implement. We can open issues about those on the retworkx repo. `dfs_edges`  should be very easy to implement since there is builtin traverser in the dag library used in retworkx: https://docs.rs/petgraph/0.5.0/petgraph/visit/struct.Dfs.html . `find_cycle` doesn't really apply to the PyDAG class because at least for right now it's checking that we don't introduce a cycle whenever an edge is added. 

You can continue to use networkx with this pr, if you want it from the dagcircuit class you can do `to_networkx()` and it will return a networkx `MultiDiGraph` object like what was there before using retworkx: https://github.com/Qiskit/qiskit-terra/pull/3810/files#diff-feb434f18758792899804d6e18c57707R85-R94 I added this for compatibility with the dag drawer (which uses network. It's a one way transformation right now so `DAGCircuit`->`networkx.MultiDiGraph` but there is no reverse method for going from networkx to DAGCircuit here.

**Comment by mtreinish** [comment]
Just for an idea of the speed up this provides the travis results for benchmarks show this pretty well: https://travis-ci.com/Qiskit/qiskit-terra/jobs/286130264#L1493

File: qiskit/dagcircuit/dagcircuit.py [hunk 1]
```diff
@@ -994,44 +1000,46 @@ def threeQ_or_more_gates(self):
...
         and [DAGNode] is its successors in  BFS order.
         """
-        return nx.bfs_successors(self._multi_graph, node)
+        return iter(rx.bfs_successors(self._multi_graph, node._node_id))
 
     def quantum_successors(self, node):
         """Returns iterator of the successors of a node that are
         connected by a quantum edge as DAGNodes."""
-        for successor in self.successors(node):
+        for successor in reversed(list(self.successors(node))):
```
[the remark this answers is not in the document]
Comment (mtreinish) [reply qiskit/dagcircuit/dagcircuit.py]: This was needed from the test results, I think the traversal order was different or something. But, there was a test case expecting the output in an order that was backwards from what retworkx was doing

File: qiskit/dagcircuit/dagcircuit.py [hunk 2]
```diff
@@ -262,25 +262,28 @@ def apply_operation_back(self, op, qargs=None, cargs=None, condition=None):
...
-        self._add_op_node(op, qargs, cargs, condition)
+        node_index = self._add_op_node(op, qargs, cargs, condition)
 
         # Add new in-edges from predecessors of the output nodes to the
         # operation node while deleting the old in-edges of the output nodes
         # and adding new edges from the operation node to each output node
         al = [qargs, all_cbits]
         for q in itertools.chain(*al):
-            ie = list(self._multi_graph.predecessors(self.output_map[q]))
+            ie = list(self._multi_graph.predecessors(self.output_map[q]._node_id))
```
[the remark this answers is not in the document]
Comment (mtreinish) [reply qiskit/dagcircuit/dagcircuit.py]: I like that idea. We rely more on node id now because retworkx is index addressed so having a public attribute with less characters would be nice. Were you thinking of a property/setter pair?

[the remark this answers is not in the document]
Comment (itoko) [reply qiskit/dagcircuit/dagcircuit.py]: I was just forgetting about setter (`node._node_id = 123` was in mind) but providing setter would be nice. I have no preference. It's up to you.

[the remark this answers is not in the document]
Comment (kdk) [reply qiskit/dagcircuit/dagcircuit.py]: In my opinion, `node_id` shouldn't be part of the `DAGNode` API and in the long term, shouldn't be tracked by `DAGNode` instances. The prior implementation had taken steps toward removing `node_id` as it had led to confusion (incorrect use as a proxy for things like graph ordering, instances where a `DAGNode` is a member of more than one `DAGCircuit`).

Instead, `DAGCircuit` could track a bidirectional `[Int, DAGNode]` map. This doesn't need to happen here, but I'd advocate against making `node_id` public.

File: qiskit/dagcircuit/dagcircuit.py [hunk 3]
```diff
@@ -702,14 +705,9 @@ def __eq__(self, other):
...
         oth = copy.deepcopy(other._multi_graph)
-
-        for node in slf.nodes:
-            slf.nodes[node]['node'] = node
-        for node in oth.nodes:
-            oth.nodes[node]['node'] = node
-
-        return nx.is_isomorphic(slf, oth,
-                                node_match=lambda x, y: DAGNode.semantic_eq(x['node'], y['node']))
+        return rx.is_isomorphic_node_match(slf, oth,
```
Comment (kdk) [inline qiskit/dagcircuit/dagcircuit.py]: Maybe the deep copies (and the comment :) ) can be removed here. They were needed because `nx.is_isomorphic` provides `node_match` with the node attributes and not the graph nodes, so we annotated each graph node with a reference to itself. I'm not sure if retworkx follows the same behavior.

File: qiskit/dagcircuit/dagcircuit.py [hunk 4]
```diff
@@ -920,25 +920,31 @@ def node(self, node_id):
...
         """
-        for source_node, dest_node, edge_data in self._multi_graph.edges(nodes, data=True):
-            yield source_node, dest_node, edge_data
+        if nodes is None:
+            nodes = self._multi_graph.nodes()
+
+        elif isinstance(nodes, DAGNode):
+            nodes = [nodes]
+        for node in nodes:
+            for source, dest, edge in self._multi_graph.in_edges(node._node_id):
```
[the remark this answers is not in the document]
Comment (kdk) [reply qiskit/dagcircuit/dagcircuit.py]: Agree we should update the docstring. On a valid `DAGCircuit`, for nodes of type `op`, `in_edges == out_edges`, but for `in`/`out` nodes, they would differ.

File: qiskit/dagcircuit/dagcircuit.py [hunk 5]
```diff
@@ -1269,8 +1278,17 @@ def nodes_on_wire(self, wire, only_ops=False):
                 yield current_node
 
             # find the adjacent node that takes the wire being looked at as input
-            for node, edges in self._multi_graph.adj[current_node].items():
-                if any(wire == edge['wire'] for edge in edges.values()):
+            # TODO(mtreinish): Add function in retworkx that does this nested api
```
Comment (kdk) [inline qiskit/dagcircuit/dagcircuit.py]: Is this fixed with `.adj_direction(direction=False)`?

Comment (mtreinish) [reply qiskit/dagcircuit/dagcircuit.py]: Unfortunately it doesn't because of multi graphs, `adj_direction` uses petgraph's `find_edge` method currently (https://docs.rs/petgraph/0.5.0/petgraph/stable_graph/struct.StableGraph.html#method.find_edge ) Which only returns a single edge. This needs to use all the edges which is why it loops over the adjacent nodes and calls `get_all_edge_data` to get the value of all the edges, even for multigraphs.

That being said @gadial has a PR open starting to work on this here: https://github.com/mtreinish/retworkx/pull/25 I haven't had a chance to review it in depth yet though.

File: qiskit/dagcircuit/dagcircuit.py (hunk 3 above)
Comment (mtreinish) [reply qiskit/dagcircuit/dagcircuit.py]: Heh, I agree the comment is probably not needed anymore. But, when I tested this without the deepcopies I got some weird hangs/performance issues running tests locally. I can try it again without them, since I added the deepcopy calls back prior to retworkx 0.2.0

File: qiskit/dagcircuit/dagcircuit.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (mtreinish) [reply qiskit/dagcircuit/dagcircuit.py]: That's fine, I don't really have a strong opinion on it. I'll leave it as is for now. FWIW, when I first started this PR I had `_id_to_node` dict and a `_node_to_id` dict. But, it got to be pretty verbose adding things like `self._multi_graph.predecessors(self._node_to_id[self.output_map[q]])` all over the place, which is why I settled on using the `_node_id` attribute in the dagnode objects

File: qiskit/dagcircuit/dagcircuit.py [hunk 6]
```diff
@@ -994,44 +1000,46 @@ def threeQ_or_more_gates(self):
...
 
     def quantum_predecessors(self, node):
         """Returns iterator of the predecessors of a node that are
         connected by a quantum edge as DAGNodes."""
-        for predecessor in self.predecessors(node):
-            if isinstance(self._multi_graph.get_edge_data(predecessor, node, key=0)['wire'], Qubit):
+        for predecessor in reversed(list(self.predecessors(node))):
+            if isinstance(
+                    self._multi_graph.get_edge_data(
+                        predecessor._node_id, node._node_id)['wire'], Qubit):
```
[the remark this answers is not in the document]
Comment (mtreinish) [reply qiskit/dagcircuit/dagcircuit.py]: Oh, that's a good catch. It's a bug in this function in it's use of the retworkx api. (for the same multigraph thing I mentioned on another comment) This should be `self._multi_graph.get_all_edge_data` and check all instances in the list for a qubit wire.

File: qiskit/dagcircuit/dagcircuit.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (mtreinish) [reply qiskit/dagcircuit/dagcircuit.py]: I just removed this and reran the tests locally only one test is failing:

```
test.python.test_dagcircuit.TestDagOperations.test_quantum_successors
---------------------------------------------------------------------

Captured traceback:
~~~~~~~~~~~~~~~~~~~
    Traceback (most recent call last):
      File "/home/mtreinish/git/qiskit/qiskit-core/test/python/test_dagcircuit.py", line 378, in test_quantum_successors
        self.assertEqual(next(successor_cnot).type, 'out')
      File "/usr/lib64/python3.8/unittest/case.py", line 912, in assertEqual
        assertion_func(first, second, msg=msg)
      File "/usr/lib64/python3.8/unittest/case.py", line 1292, in assertMultiLineEqual
        self.fail(self._formatMessage(msg, standardMsg))
      File "/usr/lib64/python3.8/unittest/case.py", line 753, in fail
        raise self.failureException(msg)
    AssertionError: 'op' != 'out'
    - op
    + out
```

File: qiskit/dagcircuit/dagcircuit.py [hunk 7]
```diff
@@ -138,34 +145,24 @@ def _add_wire(self, wire):
...
 
             wire_name = "%s[%s]" % (wire.register.name, wire.index)
 
-            inp_node = DAGNode(data_dict={'type': 'in', 'name': wire_name, 'wire': wire},
-                               nid=input_map_wire)
-            outp_node = DAGNode(data_dict={'type': 'out', 'name': wire_name, 'wire': wire},
-                                nid=output_map_wire)
+            inp_node = DAGNode(data_dict={'type': 'in', 'name': wire_name, 'wire': wire})
+            outp_node = DAGNode(data_dict={'type': 'out', 'name': wire_name, 'wire': wire})
+            input_map_wire = self._multi_graph.add_node(inp_node)
```
[the remark this answers is not in the document]
Comment (mtreinish) [reply qiskit/dagcircuit/dagcircuit.py]: Done in: https://github.com/Qiskit/qiskit-terra/pull/3810/commits/c966f02402e835021c20ae728c82c9e16c5ccfb4

File: qiskit/dagcircuit/dagcircuit.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (mtreinish) [reply qiskit/dagcircuit/dagcircuit.py]: Fixed in: https://github.com/Qiskit/qiskit-terra/pull/3810/commits/df5716141e879e74398aac6f46ee5b63d44e1bea

File: qiskit/dagcircuit/dagcircuit.py (hunk 3 above)
Comment (mtreinish) [reply qiskit/dagcircuit/dagcircuit.py]: I just tested this locally without deep copies on the current version of the branch and it both slows down test execution (mostly because of `test_examples`) and causes a panic: `thread '<unnamed>' panicked at 'index out of bounds: the len is 4 but the index is 4'` in the isomorphism function. This is because of a bug in retworkx's isomorphism check function which I just copy and pasted form petgraph and changed the input type. It's failing to deal with comparing a `StableGraph` which has a monotonically increasing index (vs a regular `Graph` which will reuse an index from a removed node). Having the deepcopy works around this issue because a deepcopy resets the index counts when it reconstructs the graph (see: https://github.com/mtreinish/retworkx/blob/master/src/lib.rs#L259-L294 )

Anyway I've added an issue about this on retworkx: https://github.com/mtreinish/retworkx/issues/27 and I'll update the comment here to reflect why it's needed.

**Comment by mtreinish** [comment]
> I'll have to look into those algorithms to see how they work, but none of the algorithms I've looked at so far have been difficult to implement. We can open issues about those on the retworkx repo. `dfs_edges` should be very easy to implement since there is builtin traverser in the dag library used in retworkx: https://docs.rs/petgraph/0.5.0/petgraph/visit/struct.Dfs.html . `find_cycle` doesn't really apply to the PyDAG class because at least for right now it's checking that we don't introduce a cycle whenever an edge is added.
> 
> You can continue to use networkx with this pr, if you want it from the dagcircuit class you can do `to_networkx()` and it will return a networkx `MultiDiGraph` object like what was there before using retworkx: https://github.com/Qiskit/qiskit-terra/pull/3810/files#diff-feb434f18758792899804d6e18c57707R85-R94 I added this for compatibility with the dag drawer (which uses networkx). It's a one way transformation right now so `DAGCircuit`->`networkx.MultiDiGraph` but there is no reverse method for going from networkx to DAGCircuit here.

I recently pushed out retworkx 0.3.0 to speed things up further by defaulting off the cycling checking, and adding a function to implement `multigraph_layers()`. Also, as part of that release @lcapelluto wrote an implementation of `floyd_warshall` for retworkx too (https://retworkx.readthedocs.io/en/stable/api.html and https://github.com/mtreinish/retworkx/pull/26). 

However, looking at the 2 PRs adding new passes it looks like they're constructing additional directed and undirected graphs for tracking the coupling map and other aspects of the algorithms being added which are outside of the dagcircuit class. Right now retworkx only has `PyDAG` class which was written assuming being used by dagcircuit. However, in the future we're planning to add types for a generic directed and undirected graph (https://github.com/mtreinish/retworkx/issues/20 ) which could be used by the passes in those PRs once implemented. I was just prioritizing the dagcircuit to start because it has the bigger impact on performance.

**Comment by ajavadia** [comment]
Ok #3762 is going to merge soon, can you rebase after that? Hopefully not much needs changing. Then I'll take another look

**Comment by mtreinish** [comment]
> Ok #3762 is going to merge soon, can you rebase after that? Hopefully not much needs changing. Then I'll take another look

Taking a quick look at #3762 I don't think there will be any changes needed, since it only interacts with the dagcircuit api which doesn't change here, but I'll rebase it after it merges and keep an eye on it.

**Comment by mtreinish** [comment]
I just ran the tests locally after rebasing and everything worked fine without any changes

File: qiskit/dagcircuit/dagcircuit.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (kdk) [reply qiskit/dagcircuit/dagcircuit.py]: I think the failure here is related to @itoko's next comment on the using out_edges vs in_edges for the definition of `edges`.

File: test/python/test_dagcircuit.py [hunk 8]
```diff
@@ -565,7 +565,7 @@ def test_remove_non_op_node(self):
         """Try to remove a non-op node with remove_op_node method."""
         self.dag.apply_operation_back(HGate(), [self.qubit0])
 
-        in_node = next(self.dag.topological_nodes())
+        in_node = next(iter(self.dag.topological_nodes()))
```
Comment (kdk) [inline test/python/test_dagcircuit.py]: I don't think the `iter` is needed anymore.

File: qiskit/dagcircuit/dagcircuit.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (mtreinish) [reply qiskit/dagcircuit/dagcircuit.py]: Hmm, I tried both updating the `edges` function and removing the `reversed` here individually and both introduced failures by themselves. But let me give fixing that func and removing this together and see if I can get that working

File: qiskit/dagcircuit/dagcircuit.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (mtreinish) [reply qiskit/dagcircuit/dagcircuit.py]: I will switch this to `out_edges` so it's consistent with what we got before. But, I just tested it locally to figure out why it didn't cause any failures. It doesn't really matter because of the structure of the dagcircuit as @kdk pointed out, since input nodes have no parents and output nodes have no children we still return all the edges for the case when nodes is None. But it could matter when nodes are set. But, I only found one place in the code base where `edges()` is called with the node params set, in the commutation analysis pass where it runs:

```
for node in dag.topological_op_nodes:
    dag.edges(node)
```

so we never encounter that edge case in the tests. When I switch to `out_edges` I'll add a test case for it.

[the remark this answers is not in the document]
Comment (mtreinish) [reply qiskit/dagcircuit/dagcircuit.py]: Done in: https://github.com/Qiskit/qiskit-terra/pull/3810/commits/5f46cc5a21c58be28a71b002b66a49f62d21dbda

File: test/python/test_dagcircuit.py (hunk 8 above)
Comment (mtreinish) [reply test/python/test_dagcircuit.py]: Done in: https://github.com/Qiskit/qiskit-terra/pull/3810/commits/af1905120381afdee3597bebe09769a1ab51d334

File: qiskit/dagcircuit/dagcircuit.py [hunk 9]
```diff
@@ -1229,14 +1238,14 @@ def collect_runs(self, namelist):
                     and not nodes_seen[node]:
                 group = [node]
                 nodes_seen[node] = True
-                s = list(self._multi_graph.successors(node))
+                s = self._multi_graph.successors(node._node_id)
```
[the remark this answers is not in the document]
Comment (mtreinish) [reply qiskit/dagcircuit/dagcircuit.py]: I tried to keep it the same (and slipped up here) to minimize the number of lines changed in order to make reverting this for any reason before the release easier. (my goal with retworkx was to try and make it as close to a drop-in `s/nx/rx/g` as possible. But, I agree we should be consistent here. Looking at the diff locally removing all the `list` casts it's only 5 more lines. So I'll just push that up.

[the remark this answers is not in the document]
Comment (mtreinish) [reply qiskit/dagcircuit/dagcircuit.py]: Done in: https://github.com/Qiskit/qiskit-terra/pull/3810/commits/c67dc50fd9121d3384842769dc3a50dffee33e66

**Comment by mtreinish** [comment]
After our discussion earlier today I've been thinking more about how we can provide a fallback for the next release in case something unforeseen comes up. Besides merging this sooner so we have time to test and debug any issues that arise, I was thinking  maybe adding a copy of the existing networkx DAGCircuit class and freezing changes to DAGCircuit's api for the rest of the release. Then when 0.13.0 is released we delete that fallback copy. So instead of having to keep up with changes and having a dual maintenance problem we just agree to not change the api until after the release (or an alternative abstraction layer on top of dagcircuit). Then if a user encounters an unforeseen issue with the new library they can just monkey patch `circuit_to_dag` and `dag_to_circuit` to use the fallback class until we get a bugfix release out. Thoughts @kdk @1ucian0 @ajavadia @ewinston ?

**Comment by kdk** [comment]
I'm on board with this idea. Two comments:
- It would be better to have the networkx/retworkx toggle set by say an environment variable. Monkey patching the converters might not be entirely straightforward for users, and requires restarting a session. (We would have to check that the env variable propagates through e.g.  `parallel_map` across platforms.)

- For the rollout, I'd vote for having 1) announce retworkx as opt-in, 2) default retworkx+deprecate networkx, and 3) remove networkx as separate steps, at one release each. This puts the process relatively on par with our standard deprecation policy and allows some time for issues to arise and be reported.

For the dual maintenance issue, I don't expect it will be a huge problem, `DAGCircuit` doesn't generally see a ton of activity: https://github.com/Qiskit/qiskit-terra/commits/master/qiskit/dagcircuit/dagcircuit.py . CI coverage would be more difficult though. Maybe convert current CI jobs to retworkx and add one compatibility job for networkx.

**Comment by mtreinish** [comment]
> I'm on board with this idea. Two comments:
> 
>     * It would be better to have the networkx/retworkx toggle set by say an environment variable. Monkey patching the converters might not be entirely straightforward for users, and requires restarting a session. (We would have to check that the env variable propagates through e.g.  `parallel_map` across platforms.)

Sure we can do this, monkey patching just seemed the easiest to me if we documented it in the release note (since it wouldn't require adding anything to the code).

> 
>     * For the rollout, I'd vote for having 1) announce retworkx as opt-in, 2) default retworkx+deprecate networkx, and 3) remove networkx as separate steps, at one release each. This puts the process relatively on par with our standard deprecation policy and allows some time for issues to arise and be reported.

I still think making it opt-out is a better approach here. This isn't really a deprecation it's a new feature (significantly improved performance) and we don't normally go through such a conservative process for new features. There will always be bugs (and often are in new features), but I agree that since this is such a core piece of terra being a bit more conservative makes sense. So I'd rather we give users a relief valve in case we're missing something and they encounter a bug. So we don't push out a release that breaks someone with no alternative. If we make it opt-in I doubt most users will ever see it until we flip the switch for real and then we're just delaying finding any potential bugs we're missing until that future date. 

> 
> 
> For the dual maintenance issue, I don't expect it will be a huge problem, `DAGCircuit` doesn't generally see a ton of activity: https://github.com/Qiskit/qiskit-terra/commits/master/qiskit/dagcircuit/dagcircuit.py . CI coverage would be more difficult though. Maybe convert current CI jobs to retworkx and add one compatibility job for networkx.

The reason I'm concerned about dual maintenance is that there are at least 2 or 3 PRs open right now besides this one which touch the DAGCircuit API: https://github.com/Qiskit/qiskit-terra/pull/3772, https://github.com/Qiskit/qiskit-terra/pull/3503, and https://github.com/Qiskit/qiskit-terra/pull/3860 (or docs in the case of this one). It's true in general there aren't many changes but I'd rather just have us make sure we're agreeing to minimize changes during the period of having multiple copies. Otherwise we're basically making us duplicate all the work

**Comment by kdk** [comment]
> > ```
> > * For the rollout, I'd vote for having 1) announce retworkx as opt-in, 2) default retworkx+deprecate networkx, and 3) remove networkx as separate steps, at one release each. This puts the process relatively on par with our standard deprecation policy and allows some time for issues to arise and be reported.
> > ```
> 
> I still think making it opt-out is a better approach here. This isn't really a deprecation it's a new feature (significantly improved performance) and we don't normally go through such a conservative process for new features. There will always be bugs (and often are in new features), but I agree that since this is such a core piece of terra being a bit more conservative makes sense.

I think that's the key concern. Unlike adding a new feature, this is a restructuring that affects nearly every qiskit user. 

> So I'd rather we give users a relief valve in case we're missing something and they encounter a bug. So we don't push out a release that breaks someone with no alternative. If we make it opt-in I doubt most users will ever see it until we flip the switch for real and then we're just delaying finding any potential bugs we're missing until that future date.

The goal of a release of opt-in is increasing exposure beyond the terra development team, to the broader community and those with an active performance problem that could be addressed by retworkx.  True there will always be bugs, but a rollout helps us gauge rough scale and severity, and more importantly, to learn early in the transition if there are users or use cases for which the new structure is difficult or not viable.  From all the work on retworkx, we don't expect it will be, but there will always be unforeseen issues.

> > For the dual maintenance issue, I don't expect it will be a huge problem, `DAGCircuit` doesn't generally see a ton of activity: https://github.com/Qiskit/qiskit-terra/commits/master/qiskit/dagcircuit/dagcircuit.py . CI coverage would be more difficult though. Maybe convert current CI jobs to retworkx and add one compatibility job for networkx.
> 
> The reason I'm concerned about dual maintenance is that there are at least 2 or 3 PRs open right now besides this one which touch the DAGCircuit API: #3772, #3503, and #3860 (or docs in the case of this one). It's true in general there aren't many changes but I'd rather just have us make sure we're agreeing to minimize changes during the period of having multiple copies. Otherwise we're basically making ourselves duplicate all the work

There are a few ways we can manage dual implementations, but I wouldn't want to forgo a rollout to ease this process. (None of those three PRs are urgent.)

**Comment by kdk** [comment]
> > ```
> > * For the rollout, I'd vote for having 1) announce retworkx as opt-in, 2) default retworkx+deprecate networkx, and 3) remove networkx as separate steps, at one release each. This puts the process relatively on par with our standard deprecation policy and allows some time for issues to arise and be reported.
> > ```
> 
> I still think making it opt-out is a better approach here. This isn't really a deprecation it's a new feature (significantly improved performance) and we don't normally go through such a conservative process for new features. There will always be bugs (and often are in new features), but I agree that since this is such a core piece of terra being a bit more conservative makes sense.

I think that's the key concern. Unlike adding a new feature, this is a restructuring that affects nearly every qiskit user. 

> So I'd rather we give users a relief valve in case we're missing something and they encounter a bug. So we don't push out a release that breaks someone with no alternative. If we make it opt-in I doubt most users will ever see it until we flip the switch for real and then we're just delaying finding any potential bugs we're missing until that future date.

The goal of a release of opt-in is increasing exposure beyond the terra development team, to the broader community and those with an active performance problem that could be addressed by retworkx.  True there will always be bugs, but a rollout helps us gauge rough scale and severity, and more importantly, to learn early in the transition if there are users or use cases for which the new structure is difficult or not viable.  From all the work on retworkx, we don't expect it will be, but there will always be unforeseen issues.

> > For the dual maintenance issue, I don't expect it will be a huge problem, `DAGCircuit` doesn't generally see a ton of activity: https://github.com/Qiskit/qiskit-terra/commits/master/qiskit/dagcircuit/dagcircuit.py . CI coverage would be more difficult though. Maybe convert current CI jobs to retworkx and add one compatibility job for networkx.
> 
> The reason I'm concerned about dual maintenance is that there are at least 2 or 3 PRs open right now besides this one which touch the DAGCircuit API: #3772, #3503, and #3860 (or docs in the case of this one). It's true in general there aren't many changes but I'd rather just have us make sure we're agreeing to minimize changes during the period of having multiple copies. Otherwise we're basically making ourselves duplicate all the work

There are a few ways we can manage dual implementations, but I wouldn't want to forgo a rollout to ease this process. (None of those three PRs are urgent.)

**Comment by mtreinish** [comment]
This has been superseded by https://github.com/Qiskit/qiskit-terra/pull/3916 which makes the DAGCircuit switchable and defaults to retworkx. This will give users an option to fallback to networkx for the 0.13.0 release. So I'm closing this I won't delete the branch though to keep this as a reference.
